tests/sphere_test.cpp

```diff
@@ -0,0 +1,104 @@
+
+  Vec3 alongXOneHitOrigin = Vec3(-5, 0, 1);
+  ASSERT_FALSE(sphere.isVecInside(alongXOneHitOrigin));
+  Ray alongXOneHit(alongXOneHitOrigin, Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXOneHit, kSkipFreq, &hitData));
+  ASSERT_EQ(5, hitData.time);
+  ASSERT_EQ(Vec3(0, 0, 1), hitData.collisionPoint());
+
+  Vec3 alongXOneHitOriginMiss = Vec3(-5, 0, 1 + constants::kAccuracy);
+  ASSERT_FALSE(sphere.isVecInside(alongXOneHitOriginMiss));
+  Ray alongXOneHitMiss(alongXOneHitOriginMiss, Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(alongXOneHitMiss, kSkipFreq, &hitData));
+  ASSERT_EQ(5, hitData.time);
+  ASSERT_EQ(Vec3(0, 0, 1), hitData.collisionPoint());
+}
+
+TEST(SphereCollisionTest, RayHitInsideSphere) {
+  Sphere sphere(Vec3::kZero, 1);
+  RayHitData hitData;
+
+  Ray alongXAxis(Vec3::kZero, Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXAxis, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(sphere.getRadius(), hitData.time);
+
+  Vec3 arbitraryChosenOriginInsideSphere(0.213, 0.523, 0.123);
+  ASSERT_TRUE(sphere.isVecInside(arbitraryChosenOriginInsideSphere));
+  Ray arbitraryRayInsideSphere(arbitraryChosenOriginInsideSphere,
+                               Vec3(214.2345, 2345, 2235.456));
+  ASSERT_TRUE(sphere.hitObject(arbitraryRayInsideSphere, kSkipFreq, &hitData));
+
+  Vec3 arbitraryInsideSphere = Vec3(12, 0.345, 3.457).normalize() * 0.99;
+  ASSERT_TRUE(sphere.isVecInside(arbitraryInsideSphere));
+  Ray randomInside(arbitraryInsideSphere, Vec3::kY);
+  ASSERT_TRUE(sphere.hitObject(randomInside, kSkipFreq, &hitData));
+}
+
+TEST(SphereCollisionTest, RayAtEdgeOfSphere) {
+  Sphere sphere(Vec3::kZero, 1);
+  RayHitData hitData;
+
+  Ray alongXAxis(Vec3::kX, Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(alongXAxis, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+  Ray alongXAxisHit(Vec3(1 - constants::kAccuracy, 0, 0), Vec3::kX);
+  ASSERT_TRUE(sphere.hitObject(alongXAxisHit, kSkipFreq, &hitData));
+  ASSERT_NEAR(constants::kAccuracy, hitData.time, constants::kAccuracy / 2);
+
+  // Prevents of getting ray inside to the sphere by
+  // floating point number error
+  Ray toOriginOfSphere(Vec3::kX, -Vec3::kX);
+  ASSERT_FALSE(sphere.hitObject(toOriginOfSphere, kSkipFreq, &hitData));
+  Vec3 xOffset(constants::kAccuracy, 0, 0);
+  Ray toOriginOfSphereFromOutside(Vec3::kX + 1.1 * xOffset, -Vec3::kX);
+  ASSERT_TRUE(
+      sphere.hitObject(toOriginOfSphereFromOutside, kSkipFreq, &hitData));
+  ASSERT_NEAR(constants::kAccuracy, hitData.time, constants::kAccuracy);
+
+  Ray parpendicularToRadius(Vec3::kX, Vec3::kY);
+  ASSERT_FALSE(sphere.hitObject(parpendicularToRadius, kSkipFreq, &hitData));
+
+  Ray offsettedDirectionParpendicularToRadius(Vec3::kX, Vec3::kY - xOffset);
+  ASSERT_FALSE(sphere.hitObject(offsettedDirectionParpendicularToRadius,
+                                kSkipFreq, &hitData));
+  Ray offsettedParpendicularToRadius(Vec3::kX + xOffset,
+                                     Vec3::kY - 1000 * xOffset);
```

Comment:
How is this perpendicular? 

I am not sure what this test is checking. I thought raus perpendicular to radius would never hit.

Reply:
I was trying to test edge case, when ray almost parpendicular to radius actaully hits the sphere. Deleted it, because its not usefull actually in this case ;)